### PR TITLE
Adds ruby challenges

### PR DESCRIPTION
**Pull request by mikemar10**
Hey Yehonathan!

I came across your blog post today on https://lobste.rs and had fun trying your challenge in Ruby. I used the latest Ruby 3.0.0 for this, I imagine these examples should work with any version of Ruby since 2.4 but haven't personally confirmed.

Thanks for giving me an excuse to exercise my brain this evening :)

### CODE REVIEW

File: challenges/ruby/book-property.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+require 'pp'
+load 'library-data.rb'
+
+def get_book_property(library_data, isbn, field_name)
+  information_path = ['catalog', 'booksByIsbn', isbn, field_name]
+  library_data.dig(*information_path)
```
Comment (viebel) [inline challenges/ruby/book-property.rb:6]: Cool! I didn't know that Ruby provided natively a way to access data in a hash map via its path

File: challenges/ruby/library-data.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
+class Hash
+  def deep_copy
+    Marshal.load(Marshal.dump(self))
```
Comment (viebel) [inline challenges/ruby/library-data.rb:3]: Oh no! There must be a better way to deeply copy a hash map!

File: challenges/ruby/merge-and-serialize.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
+require 'json'
+require 'pp'
+load 'library-data.rb'
+
+def merge_and_serialize(a, b)
+  a.merge(b).to_json
```
Comment (viebel) [inline challenges/ruby/merge-and-serialize.rb:6]: So neat!

File: challenges/ruby/rename-keys.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
+require 'pp'
+load 'library-data.rb'
+
+def rename_keys(map, key_map)
+  key_map.reduce(map.deep_copy) do |result, (old_key, new_key)|
```
Comment (viebel) [inline challenges/ruby/rename-keys.rb:5]: Here a shallow copy would be enough.

File: challenges/ruby/block-member.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+require 'pp'
+load 'library-data.rb'
+
+def block_member(library_data, email)
+  information_path = ['userManagement', 'members', email, 'isBlocked']
+  library_data.deep_copy.tap do |new_library_data|
```
Comment (viebel) [inline challenges/ruby/block-member.rb:6]: deep copy works here but it causes a perf issue.
A better way would be to implement structural sharing.
It's only 7 lines of code in JavaScript: https://blog.klipse.tech/javascript/2021/02/26/structural-sharing-in-javascript.html

**Review by viebel — CHANGES_REQUESTED** [review]
Well done!
It's cool to see how easy it is to write code like that in Ruby.

File: challenges/ruby/library-data.rb (hunk 2 above)
Comment (mikemar10) [reply challenges/ruby/library-data.rb:3]: I felt the same way, I attempted first to use `clone` and `dup` but both are shallow copies. This was surprisingly the method suggested in several places online. The only other ready to go solution I've come across is `deep_dup` from [Rails ActiveSupport](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/core_ext/object/deep_dup.rb). I'm going to pull it in, I haven't used just activesupport in a project before so it'll be fun to go through that.

File: challenges/ruby/merge-and-serialize.rb (hunk 3 above)
Comment (mikemar10) [reply challenges/ruby/merge-and-serialize.rb:6]: ❤️ ruby

File: challenges/ruby/block-member.rb (hunk 5 above)
Comment (mikemar10) [reply challenges/ruby/block-member.rb]: adding structural sharing was really cool! I was loosely aware of the concept in the sense that I know clojure does something similar under the hood but I just assumed it was complicated and never looked closely. Thank you so much for sharing that, I was pleasantly surprised at how easy it was to implement and how it removed the need for deep copying altogether. I had added activesupport in to get a less hacky implementation of deep copying and I just ripped it out completely thanks to this!

File: challenges/ruby/rename-keys.rb (hunk 4 above)
Comment (mikemar10) [reply challenges/ruby/rename-keys.rb]: I switched this to a shallow copy and it changed the behavior. Troubleshot it for a bit but ended up rewriting the function using `Hash#transform_keys`

File: challenges/ruby/book-property.rb (hunk 1 above)
Comment (mikemar10) [reply challenges/ruby/book-property.rb:5]: it's somewhat new in the last couple years but pretty handy. Surprisingly there isn't a convenient update_in type function but it was pretty easy to monkeypatch it in as part of your structural sharing suggestion.

File: challenges/ruby/library-data.rb [hunk 6]
```diff
@@ -53,4 +91,4 @@ def deep_copy
       }
     }
   }
-}
+}.immutable!
```
Comment (viebel) [inline challenges/ruby/library-data.rb:94]: Yeah!!!

**Review by viebel — APPROVED** [review]

File: challenges/ruby/book-property.rb (hunk 1 above)
Comment (viebel) [reply challenges/ruby/book-property.rb:5]: Hopefully, Ruby will have `set_in` soon

File: challenges/ruby/block-member.rb (hunk 5 above)
Comment (viebel) [reply challenges/ruby/block-member.rb]: I'm so happy you liked implementing structural sharing in Ruby.
Structural sharing in Clojure is a bit more complicated as it does the structural sharing at the level of the data structure itself.
The simple implementation suffers from perf issues when there are many elements at the same level (e.g. an array with 10K elements, or a map with 10K keys).

**Comment by viebel** [comment]
@mikemar10 Could you send me a message on Twitter? You just one a free copy of the book!

**Comment by mikemar10** [comment]
@viebel thank you very much :) DM on twitter incoming

**Pull request merged**